### DISCUSSION TITLE
How to get the list of feature dataset from a FileGDB file using python

### DISCUSSION THREAD

**Discussion by kavyajeetbora** [Toblerity/Fiona#1261] [Q&A] [answered]
I want to list down the names of feature dataset. 
This is the basic structure of the geodatabase: 
![image](https://user-images.githubusercontent.com/38955297/236751416-ae31f5e6-33f4-4d15-bb36-25874631f510.png)

Using fiona I am able to get the list of feature classes using:

`layers = fiona.listlayers(gdb_filePath)`

Can anyone help me to get the list of feature datasets (not feature classes) ?
Using arcpy can be done I guess but don't have it as it needs license. Looking for some solution using opensource libraries like fiona and GDAL, if any

**Comment by sgillies** [comment, marked as answer]
@kavyajeetbora ~[fiona.listdir](https://fiona.readthedocs.io/en/stable/fiona.html#fiona.listdir) is what you're looking for. I'll update the documentation to make that more clear.~

Correction: fiona doesn't support multiple datasets within a geodatabase file. The GDAL library has some support since version 3.4 https://gdal.org/drivers/vector/openfilegdb.html#hiearchical-organization but we haven't implemented it for fiona yet. [fiona.listdir](https://fiona.readthedocs.io/en/stable/fiona.html#fiona.listdir) is only useful for finding datasets within a filesystem directory.

**Reply by kavyajeetbora** [reply]
Thanks very much @sgillies 
I got the list of datasets using this following command (using osgeo.gdal module):
`groupList = gdal.OpenEx(gdb_filePath).GetRootGroup().GetGroupNames()`

and to get the layers within a particular group:
```
dataset = gdal.OpenEx(gdb_filePath)
# GroupName from the groupList above
dataset.GetRootGroup().OpenGroup('GroupName').GetVectorLayerNames()
```
